entropy/windows/Raaz/Entropy.hsc

```diff
@@ -38,28 +35,22 @@ foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptAcquireContextW"
 foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptReleaseContext"
     c_CryptReleaseContext :: HCRYPTPROV -> Word32 -> IO Bool
 
--- | Cache the crypto context so we don't have to create it on each call.
-cryptoContext :: ForeignPtr HCRYPTPROV
-cryptoContext = unsafePerformIO $
-  do buffer <- mallocBytes (#size HCRYPTPROV)
-     addr   <- new buffer
-     ctx    <- newForeignPtr addr (freeContext buffer >> free addr)
-     ctx_ok <- withForeignPtr ctx $ \ptr ->
-                    c_CryptAcquireContext ptr nullPtr nullPtr
-                        (#const PROV_RSA_FULL)
-                        ((#const CRYPT_VERIFYCONTEXT) .|. (#const CRYPT_SILENT))
-     if ctx_ok
-        then return ctx
-        else error "Call to CryptAcquireContext failed."
-
--- | Release the crytographical context handle.
-freeContext :: HCRYPTPROV -> IO ()
-freeContext ctx = c_CryptReleaseContext ctx 0 >> return ()
-
 -- | Get cryptographically random bytes from the system.
 getEntropy :: (MonadIO m, LengthUnit l) => l -> Pointer -> m (BYTES Int)
-getEntropy l ptr = liftIO $ withForeignPtr cryptoContext $ \ctx ->
-    do ctx' <- peek ctx
+getEntropy l ptr = liftIO $
+    -- You can't get an address of a pointer in haskell, so making out
+    -- parameters are messy. We do this by boxing the type.
+    do ctx    <- mallocBytes (#size HCRYPTPROV)
```

Comment:
For the context you need a `Ptr  something` (may be `Ptr Word8`). Is that correct?  There is a combinator called [`allocaBuffer`](https://www.stackage.org/haddock/lts-8.9/raaz-0.1.1/Raaz-Core-Types.html) in the module `Raaz.Core.Types`. It is a safer (and often faster) form of malloc and free. There are variants that take care alignment restrictions as well. 

These family of allocation functions are built on top of the [`allocaBytes`](https://hackage.haskell.org/package/base-4.9.1.0/docs/Foreign-Marshal-Alloc.html)  function  from the `Foreign.Marshall.Alloc` modules.